src/torcontrol.cpp

```diff
@@ -0,0 +1,685 @@
+{
+    if (b_conn)
+        Disconnect();
+    // Parse target address:port
+    struct sockaddr_storage connect_to_addr;
+    int connect_to_addrlen = sizeof(connect_to_addr);
+    if (evutil_parse_sockaddr_port(target.c_str(),
+        (struct sockaddr*)&connect_to_addr, &connect_to_addrlen)<0) {
+        LogPrintf("tor: Error parsing socket address %s\n", target);
+        return false;
+    }
+
+    // Create a new socket, set up callbacks and enable notification bits
+    b_conn = bufferevent_socket_new(base, -1, BEV_OPT_CLOSE_ON_FREE);
+    if (!b_conn)
+        return false;
+    bufferevent_setcb(b_conn, TorControlConnection::readcb, NULL, TorControlConnection::eventcb, this);
+    bufferevent_enable(b_conn, EV_READ|EV_WRITE);
+    this->connected = connected;
+    this->disconnected = disconnected;
+
+    // Finally, connect to target
+    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&connect_to_addr, connect_to_addrlen) < 0) {
+        LogPrintf("tor: Error connecting to address %s\n", target);
+        return false;
+    }
+    return true;
+}
+
+bool TorControlConnection::Disconnect()
+{
+    if (b_conn)
+        bufferevent_free(b_conn);
+    b_conn = 0;
+    return true;
+}
+
+bool TorControlConnection::Command(const std::string &cmd, const ReplyHandlerCB& reply_handler)
+{
+    if (!b_conn)
+        return false;
+    struct evbuffer *buf = bufferevent_get_output(b_conn);
+    if (!buf)
+        return false;
+    evbuffer_add(buf, cmd.data(), cmd.size());
+    evbuffer_add(buf, "\r\n", 2);
+    reply_handlers.push_back(reply_handler);
+    return true;
+}
+
+/****** General parsing utilities ********/
+
+/* Split reply line in the form 'AUTH METHODS=...' into a type
+ * 'AUTH' and arguments 'METHODS=...'.
+ */
+static std::pair<std::string,std::string> SplitTorReplyLine(const std::string &s)
+{
+    size_t ptr=0;
+    std::string type;
+    while (ptr < s.size() && s[ptr] != ' ') {
+        type.push_back(s[ptr]);
+        ++ptr;
+    }
+    if (ptr < s.size())
+        ++ptr; // skip ' '
+    return make_pair(type, s.substr(ptr));
+}
+
+/** Parse reply arguments in the form 'METHODS=COOKIE,SAFECOOKIE COOKIEFILE=".../control_auth_cookie"'.
+ */
+static std::map<std::string,std::string> ParseTorReplyMapping(const std::string &s)
+{
+    std::map<std::string,std::string> mapping;
+    size_t ptr=0;
+    while (ptr < s.size()) {
+        std::string key, value;
+        while (ptr < s.size() && s[ptr] != '=') {
+            key.push_back(s[ptr]);
+            ++ptr;
+        }
+        if (ptr == s.size()) // unexpected end of line
+            return std::map<std::string,std::string>();
+        ++ptr; // skip '='
+        if (ptr < s.size() && s[ptr] == '"') { // Quoted string
+            ++ptr; // skip '='
+            bool escape_next = false;
+            while (ptr < s.size() && (!escape_next && s[ptr] != '"')) {
+                escape_next = (s[ptr] == '\\');
+                value.push_back(s[ptr]);
```

Comment:
Yeah, see section 2.1.1 of https://gitweb.torproject.org/torspec.git/tree/control-spec.txt